modules/auxiliary/gather/ldap_esc_vulnerable_cert_finder.rb

```diff
@@ -0,0 +1,544 @@
+
+    esc3_template_1_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
+        @vuln_cert_template_table << ['ESC3_TEMPLATE_1', entry[:cn][0], entry[:dn][0], convert_sids_to_human_readable_name(allowed_sids)]
+      end
+      parse_dacl_or_sacl(security_descriptor.sacl) if security_descriptor.sacl
+    end
+
+    esc3_template_2_raw_filter = '(&'\
+      '(objectclass=pkicertificatetemplate)'\
+      '(!(mspki-enrollment-flag:1.2.840.113556.1.4.804:=2))'\
+      '(|'\
+        '(mspki-template-schema-version=1)'\
+        '(&'\
+          '(mspki-template-schema-version>=2)'\
+          '(msPKI-RA-Application-Policies=1.3.6.1.4.1.311.20.2.1)'\
+        ')'\
+      ')'\
+      '(|'\
+        '(pkiextendedkeyusage=1.3.6.1.4.1.311.20.2.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.5.7.3.2)'\
+        '(pkiextendedkeyusage=1.3.6.1.5.2.3.4)'\
+        '(pkiextendedkeyusage=2.5.29.37.0)'\
+        '(!(pkiextendedkeyusage=*))'\
+      ')'\
+    ')'
+
+    # Find the first vulnerable types of templates, those that have the OID of the
+    # Certificate Request Agent which allows the template to be used for
+    # requesting other certificate templates on behalf of other principals.
+    attributes = ['cn', 'description', 'ntSecurityDescriptor']
+    base_prefix = 'CN=Certificate Templates,CN=Public Key Services,CN=Services,CN=Configuration'
+    esc3_template_2_entries = query_ldap_server(esc3_template_2_raw_filter, attributes, base_prefix: base_prefix)
+    fail_with(Failure::NotVulnerable, 'Could not find any ESC3 template 2 vulnerable certificate templates!') if esc3_template_2_entries.blank?
+
+    # Grab a list of certificates that contain vulnerable settings.
+    # Also print out the list of SIDs that can enroll in that server.
+
+    esc3_template_2_entries.each do |entry|
+      @vuln_certificate_template_list << entry[:cn][0]
+
+      begin
+        security_descriptor = Rex::Proto::MsDtyp::MsDtypSecurityDescriptor.read(entry[:ntsecuritydescriptor][0])
+      rescue IOError => e
+        fail_with(Failure::UnexpectedReply, "Unable to read security descriptor! Error was: #{e.message}")
+      end
+
+      flag_allowed_to_enroll, allowed_sids = parse_dacl_or_sacl(security_descriptor.dacl) if security_descriptor.dacl
+      if flag_allowed_to_enroll == true
+        @vuln_cert_template_table << ['ESC3_TEMPLATE_2', entry[:cn][0], entry[:dn][0], convert_sids_to_human_readable_name(allowed_sids)]
+      end
+      parse_dacl_or_sacl(security_descriptor.sacl) if security_descriptor.sacl
```

Comment:
Same comment *again* about this block. `flag_allowed_to_enroll` is buggy, it should just be `if flag_allowed_to_enroll` and then `parse_dacl_or_sacl` doesn't appear to do anything.

Reply:
Fixed this up with https://github.com/rapid7/metasploit-framework/pull/17122/commits/fd33cdc58e47576823a874c5f17d81b3e5660f9b which uses the new function you suggested and also fixes the aforementioned issues.